File: PA124.java
//Leetcode: 379. Design Phone Directory
//Time Complexity: get -- O(n), check --O(1) , releae -- O(1)
//Space Complexity: O(n) 
class PhoneDirectory {

    /** Initialize your data structure here
        @param maxNumbers - The maximum numbers that can be stored in the phone directory. */
    
    boolean[] numbers;
    int size;
    public PhoneDirectory(int maxNumbers) {
        size=maxNumbers;
        numbers= new boolean[maxNumbers];
    }
    
    /** Provide a number which is not assigned to anyone.
        @return - Return an available number. Return -1 if none is available. */
    public int get() {
        if(size>0 ){
            for (int i=0;i<numbers.length;i++){
                if(!numbers[i]){
                    size--;
                    numbers[i]= true;
                    return i;
                }
            }
        }
        return -1;
    }
    
    /** Check if a number is available or not. */
    public boolean check(int number) {
        return (!numbers[number]);
    }
    
    /** Recycle or release a number. */
    public void release(int number) {
        size++;
        numbers[number] = false;
    }
}

/**
 * Your PhoneDirectory object will be instantiated and called as such:
 * PhoneDirectory obj = new PhoneDirectory(maxNumbers);
 * int param_1 = obj.get();
 * boolean param_2 = obj.check(number);
 * obj.release(number);
 */

File: PA124.py
#Time Complexity - O(1)
#Space Complexity - O(1)
from collections import deque
class PhoneDirectory(object):

    def __init__(self, maxNumbers):
        """
        Initialize your data structure here
        @param maxNumbers - The maximum numbers that can be stored in the phone directory.
        :type maxNumbers: int
        """
        self.size = maxNumbers
        self.q = deque()
        self.vis = set()
        for i in range(self.size):
            self.q.append(i)

    def get(self):
        """
        Provide a number which is not assigned to anyone.
        @return - Return an available number. Return -1 if none is available.
        :rtype: int
        """
        if self.size > 0:
            self.size = self.size - 1
            num = self.q.popleft()
            self.vis.add(num)
            return num
        return -1
        

    def check(self, number):
        """
        Check if a number is available or not.
        :type number: int
        :rtype: bool
        """
        return number not in self.vis

    def release(self, number):
        """
        Recycle or release a number.
        :type number: int
        :rtype: None
        """
        if number in self.vis:
            self.vis.remove(number)
            self.q.append(number)
            self.size = self.size + 1


# Your PhoneDirectory object will be instantiated and called as such:
# obj = PhoneDirectory(maxNumbers)
# param_1 = obj.get()
# param_2 = obj.check(number)
# obj.release(number)